botocore/regions.py

```diff
@@ -562,6 +575,71 @@ def _get_provider_params(
 
         return provider_params
 
+    def _resolve_account_id_builtin(self, request_context, builtins):
+        """Resolve the ``AWS::Auth::AccountId`` builtin if configured in the
+        ruleset, account ID based routing is enabled and it has not already
+        been resolved in a custom handler.
+        """
+        if 'AccountId' in self._param_definitions:
+            act_id_ep_mode = self._resolve_account_id_endpoint_mode(
+                request_context
+            )
+            act_id_builtin_key = EndpointResolverBuiltins.AWS_ACCOUNT_ID
+            if act_id_ep_mode == 'disabled':
+                # if account ID has been set with a custom handler, but the mode
+                # is disabled, we must unset it so it won't be passed to the
+                # endpoint provider.
+                builtins[act_id_builtin_key] = None
+                return
+
+            act_id_builtin = builtins.get(act_id_builtin_key)
+            if act_id_builtin is None:
+                self._do_resolve_account_id_builtin(act_id_ep_mode, builtins)
+
+    def _resolve_account_id_endpoint_mode(self, request_context):
+        """Resolve the account ID endpoint mode for the request. Account ID
+        based routing is always disabled for presigned and unsigned requests.
+        Otherwise, the mode is determined by the ``account_id_endpoint_mode``
+        config setting.
+        """
+        not_presign = not request_context.get('is_presign_request', False)
+        should_sign = self._requested_auth_scheme != UNSIGNED
+        creds_available = self._credentials is not None
+        if all((not_presign, should_sign, creds_available)):
+            config = request_context['client_config']
+            act_id_ep_mode = config.account_id_endpoint_mode
+            return self._validate_account_id_endpoint_mode(act_id_ep_mode)
+        return 'disabled'
+
+    def _validate_account_id_endpoint_mode(self, account_id_endpoint_mode):
+        if account_id_endpoint_mode not in VALID_ACCOUNT_ID_ENDPOINT_MODES:
+            valid_modes_str = ', '.join(VALID_ACCOUNT_ID_ENDPOINT_MODES)
+            error_msg = (
+                f"Invalid value '{account_id_endpoint_mode}' for "
+                "account_id_endpoint_mode. Valid values are: "
+                f"{valid_modes_str}."
+            )
+            raise InvalidConfigError(error_msg=error_msg)
+        return account_id_endpoint_mode
+
+    def _do_resolve_account_id_builtin(
+        self, account_id_endpoint_mode, builtins
+    ):
+        # This will make a call to resolve credentials if they are not already
+        # or need to be refreshed.
+        frozen_creds = self._credentials.get_frozen_credentials()
+        account_id = frozen_creds.account_id
+        if account_id is None:
+            if account_id_endpoint_mode == 'preferred':
+                LOG.debug(
```

Comment:
There was a concern about blowing up logs. I'm not opposed to changing this though

Reply:
This will be _very_ noisy for the average user who likely doesn't want/need this feature. The default is preferred which means every call will log a warning.